File: SearchA2DMatrix/Solution.rb
def search(row, target)
    l = 0
    r = row.length
    while l < r do
        m = l + (r - l) / 2
        if row[m] < target then
            l = m + 1
        elsif row[m] == target then
            return true
        else
            r = m
        end
    end
    return false
end

# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
def search_matrix(matrix, target)
    l = 0
    r = matrix.length
    while l < r do
        m = l + (r - l) / 2
        if matrix[m][0] <= target then
            l = m + 1
        else
            r = m
        end
    end
    return false if l == 0
    return search(matrix[l - 1], target)
end

File: SearchA2DMatrix/Solution.c
int search(int** matrix, int row, int target) {
    int l = 0;
    int r = row;
    int m = 0;
    while (l < r) {
        m = l + (r - l) / 2;
        if (matrix[m][0] <= target) {
            l = m + 1;
        } else {
            r = m;
        }
    }
    return r - 1;
}

bool searchRow(int* row, int col, int target) {
    int l = 0;
    int r = col;
    int m = 0;
    while (l < r) {
        m = l + (r - l) / 2;
        if (row[m] < target) {
            l = m + 1;
        } else if (row[m] == target) {
            return true;
        } else {
            r = m;
        }
    }
    return false;
}

bool searchMatrix(int** matrix, int matrixRowSize, int matrixColSize, int target) {
    if (matrixRowSize == 0) return false;
    int i = search(matrix, matrixRowSize, target);
    if (i < 0) return false;
    return searchRow(matrix[i], matrixColSize, target);
}
